README.md
# Morse Passwd For Web

[中文文档](./README.CN.md)

This package is used to set a morse passwd in pc or mobile web page.
When the user input the morse code by click and press the page, the success callback will be trigger.
It can be used to unlock a page, verify if the user is human beings, or to open a debug page.

## UsAge

Install the package from npm:

```
$ npm install morse-passwd --save;
```

Use it in web page.

```js
import MorsePass from 'morse-passwd'

moresePass = new MorsePass({
    // the target element which will bind event
    target: document.querySelector('body'),
    // the passwd to verify. '.' represents a click, '_' represents a long press
    passwd: '.._.._',
    // the time from mousedown to mouse up. it will be treated as a press if bigger than this time
    // or it will be treated as click.  the unit is ms
    pressLeastTime: 1000,
    // how long time will the current operation expired, then it will verify again from the first code. the unit is ms
    expires: 6000,
    // trigger when the passwd is right.
    success() {
        console.log('Passwd Right!')
    },
    // trigger when input wrong code
    fail() {
        console.log('Passwd Wrong!')
    },
    // trigger on each input
    input(text) {
        console.log(text)
    },
    // trigger when the time is expired after the last operation
    expire() {
        console.log('expired!')
    }
})
```

## Example

- [./examples/index.html](https://xu3927.github.io/morse-passwd/examples/index.html)
- [./examples/unlock-screen.html](https://xu3927.github.io/morse-passwd/examples/unlock-screen.html)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/xu3927/morse-passwd/compare/v1.0.4...v1.1.0) (2022-07-23)


### Features

* use touchstart,touchend for mobile; ([df8ea96](https://github.com/xu3927/morse-passwd/commit/df8ea9644a9f7ad732ba55350ac7d80e437984af))

### [1.0.4](https://github.com/xu3927/morse-passwd/compare/v1.0.3...v1.0.4) (2022-07-22)

### [1.0.3](https://github.com/xu3927/morse-passwd/compare/v1.0.1...v1.0.3) (2022-07-22)

publish to npm

### 1.0.1 (2022-07-22)

init

src/index.ts
type TConfig = {
    target?: HTMLElement | string;
    passwd: string;
    pressLeastTime?: number;
    expires?: number;
    success?(): void;
    fail?(): void;
    expire?(): void;
    input?(inp: string): void;
}

enum TYPE_MAP {
    DOT = 'dot',
    DASH = 'dash'
}
const TYPE_VALUE_MAP = {
    [TYPE_MAP.DOT]: '.',
    [TYPE_MAP.DASH]: '_'
}

const defaultConfig: TConfig = {
    // the target element which will bind event
    target: window.document.querySelector('body') as HTMLElement,
    // the passwd. '.' represents one click, '_' represents one press
    passwd: '.._.._',
    // how long time press should be counted. the unit is ms
    pressLeastTime: 1000,
    // how long time will the current operation expired, then you should do it from the first code again. the unit is ms
    expires: 1000 * 6,
    // the callback if the passwd is right.
    success() {
        console.log('Passwd Right!')
    },
    fail() {
        console.log('Passwd Wrong!')
    },
    // the callback if the operation is expired.
    expire() {
        console.log('The operation is expired, please do it again.')
    }
}


class InnerMorsePass {
    constructor(config: TConfig) {
        this.config = Object.assign({}, defaultConfig, config)
        let element: any = config.target
        if (typeof config.target === 'string') {
            element = document.querySelector(config.target)
        }
        if (!element) {
            throw new Error('the target element was ' + element)
        }
        this.element = element;
        if (typeof this.config.pressLeastTime === 'number') {
            this.pressLeastTime = this.config.pressLeastTime
        }
        if (typeof this.config.expires === 'number') {
            this.expires = this.config.expires
        }
        if (!/^[._]+$/.test(this.config.passwd)) {
            throw new Error('the passwd should only contain . and _');
        }
        this.passwd = this.config.passwd;
        this.init()
    }
    element: HTMLElement = document.querySelector('body') as HTMLElement
    config: TConfig
    pressLeastTime = 1500
    passwd = '.._..'
    steps = ['.', '.', '_', '.', '.']
    mousedownTime = Number.MAX_SAFE_INTEGER;
    expireTimer = -1;
    expires = 6000;
    init() {
        this.reset()
        this.bindListeners()
    }
    clickHandler = () => {
        this.run(TYPE_MAP.DOT)
    }
    mousedownHandler = () => {
        this.mousedownTime = new Date().getTime();
    }
    mouseupHandler = () => {
        if (new Date().getTime() - this.mousedownTime > this.pressLeastTime) {
            this.run(TYPE_MAP.DASH)
        } else {
            this.run(TYPE_MAP.DOT)
        }
    }
    bindListeners() {
        this.element.addEventListener('mousedown', this.mousedownHandler)
        this.element.addEventListener('mouseup', this.mouseupHandler)
    }
    removeListeners() {
        this.element.removeEventListener('mousedown', this.mousedownHandler)
        this.element.removeEventListener('mouseup', this.mouseupHandler)
    }
    run(type: `${TYPE_MAP}`) {
        window.clearTimeout(this.expireTimer)
        this.expireTimer = window.setTimeout(() => {
            this.expire()
        }, this.config.expires);

        if (this.steps[0] === TYPE_VALUE_MAP[type]) {
            this.next()
        } else {
            this.fail()
        }
        this.input()
    }
    next() {
        this.steps.shift()
        if (this.steps.length === 0) {
            this.success()
        }
    }
    reset() {
        window.clearTimeout(this.expireTimer)
        this.steps = this.passwd.split('')
    }
    success() {
        window.clearTimeout(this.expireTimer)
        if (typeof this.config.success === 'function') {
            this.config.success()
        }
        this.removeListeners()
    }
    fail() {
        window.clearTimeout(this.expireTimer)
        if (typeof this.config.fail === 'function') {
            this.config.fail()
        }
    }
    input() {
        if (typeof this.config.input === 'function') {
            this.config.input(this.passwd.slice(0, this.passwd.length - this.steps.length))
        }
    }
    destroy() {
        window.clearTimeout(this.expireTimer)
        this.removeListeners()
    }
    expire() {
        this.fail()
        this.reset()
        if (typeof this.config.expire === 'function') {
            this.config.expire()
        }
    }
}

class MorsePass {
    constructor(config: TConfig) {
        this.morsePass = new InnerMorsePass(config);
    }
    morsePass: InnerMorsePass
    destroy() {
        this.morsePass.destroy()
    }
}

export default MorsePass;
